File: Flutter/lab2/lib/entities/city_entity.dart
class CityEntity {
  late String id;
  late String name;
  late String country;
  late String description;
  late int yearOfFoundation;
  late List<String> imageUrls;

  CityEntity(this.id, this.name, this.country, this.description,
      this.yearOfFoundation, this.imageUrls);

  Map<String, dynamic> toJson() => {
        "id": id,
        "name": name,
        "description": description,
        "country": country,
        "yearOfFoundation": yearOfFoundation,
        "imageUrls": imageUrls
      };

  static getCityList(List data) {
    List<CityEntity> result = [];
    for (final city in data) {
      List<String> imageUrls =
          (city['imageUrls'] as List).map((e) => e as String).toList();
      result.add(CityEntity(city['id'], city['name'], city['description'],
          city['country'], city['yearOfFoundation'], imageUrls));
    }
    return result;
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;

    if (other is CityEntity) {
      return other.id == id;
    } else if (other is Map<Object?, Object?>) {
      Map<String, dynamic> temp =
          other.map((key, value) => MapEntry(key.toString(), value));
      return temp['id'] == id;
    }

    return false;
  }

  @override
  int get hashCode => name.hashCode ^ country.hashCode;
}

File: Swift/lab3/lab3/Entities/CityEntity.swift
//
//  CityEntity.swift
//  lab3
//
//  Created by dumonten on 15/3/24.
//

import Foundation

struct CityEntity: Codable {
    var id: String
    var name: String
    var description: String
    var yearOfFoundation: Int
    var country: String
    var imageUrls: [String]
    
    init() {
        self.id = ""
        self.name = ""
        self.description = ""
        self.yearOfFoundation = 0
        self.country = ""
        self.imageUrls = []
    }
    
    init(id: String, name: String, description: String, yearOfFoundation: Int, country: String, imageUrls: [String]) {
        self.id = id
        self.name = name
        self.description = description
        self.yearOfFoundation = yearOfFoundation
        self.country = country
        self.imageUrls = imageUrls
    }
    
    func toJson() -> [String: Any] {
        let encoder = JSONEncoder()
        encoder.keyEncodingStrategy = .convertToSnakeCase
        if let jsonData = try? encoder.encode(self),
           let json = try? JSONSerialization.jsonObject(with: jsonData, options: []) as? [String: Any] {
            return json
        }
        return [:]
    }
    
    static func fromJson(json: [String: Any]) -> CityEntity? {
        let decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        if let jsonData = try? JSONSerialization.data(withJSONObject: json, options: []),
           let city = try? decoder.decode(CityEntity.self, from: jsonData) {
            return city
        }
        return nil
    }
}
